pkg/store/pipelineconfig/store.go

```diff
@@ -0,0 +1,238 @@
+	}
+	return entry.Value, nil
+}
+
+func (s *configurationStore) Create(ctx context.Context, pipelineConfig *p4rtapi.PipelineConfig) error {
+	if pipelineConfig.ID == "" {
+		return errors.NewInvalid("no pipeline pipelineconfig ID specified")
+	}
+	if pipelineConfig.TargetID == "" {
+		return errors.NewInvalid("no target ID specified")
+	}
+	if pipelineConfig.Revision != 0 {
+		return errors.NewInvalid("cannot create pipeline pipeline config with revision")
+	}
+	if pipelineConfig.Version != 0 {
+		return errors.NewInvalid("cannot create pipeline pipeline config with version")
+	}
+	pipelineConfig.Revision = 1
+	pipelineConfig.Created = time.Now()
+	pipelineConfig.Updated = time.Now()
+	// Create the entry in the underlying map primitive.
+	entry, err := s.pipelineConfigs.Put(ctx, pipelineConfig.ID, pipelineConfig)
+	if err != nil {
+		return errors.FromAtomix(err)
+	}
+
+	// Decode the pipeline config from the returned entry.
+	if err := decodePipelineConfiguration(entry, pipelineConfig); err != nil {
+		return errors.NewInvalid("pipelineConfig decoding failed: %v", err)
+	}
+	return nil
+}
+
+func (s *configurationStore) Update(ctx context.Context, pipelineConfig *p4rtapi.PipelineConfig) error {
+	if pipelineConfig.ID == "" {
+		return errors.NewInvalid("no pipelineConfig ID specified")
+	}
+	if pipelineConfig.TargetID == "" {
+		return errors.NewInvalid("no target ID specified")
+	}
+	if pipelineConfig.Revision == 0 {
+		return errors.NewInvalid("pipelineConfig must contain a revision on update")
+	}
+	if pipelineConfig.Version == 0 {
+		return errors.NewInvalid("pipelineconfig must contain a version on update")
+	}
+	pipelineConfig.Revision++
+	pipelineConfig.Updated = time.Now()
+
+	// Update the entry in the underlying map primitive using the pipelineconfig version
+	// as an optimistic lock.
+	entry, err := s.pipelineConfigs.Put(ctx, pipelineConfig.ID, pipelineConfig)
+	if err != nil {
+		return errors.FromAtomix(err)
+	}
+
+	// Decode the pipelineconfig from the returned entry bytes.
+	if err := decodePipelineConfiguration(entry, pipelineConfig); err != nil {
+		return errors.NewInvalid("pipeline config decoding failed: %v", err)
+	}
+
+	return nil
+}
+
+func (s *configurationStore) List(ctx context.Context) ([]*p4rtapi.PipelineConfig, error) {
+	entryStream, err := s.pipelineConfigs.List(ctx)
+	if err != nil {
+		return nil, errors.FromAtomix(err)
+	}
+
+	var entries []*p4rtapi.PipelineConfig
+
+	for {
+		entry, err := entryStream.Next()
+		if err == io.EOF {
+			log.Warn("Entry stream is closed")
+			break
+		}
+		entries = append(entries, entry.Value)
+	}
+
+	return entries, nil
+
+}
+
+func (s *configurationStore) Watch(ctx context.Context, ch chan<- *p4rtapi.PipelineConfig, opts ...WatchOption) error {
+	entryStream, err := s.pipelineConfigs.Watch(ctx)
+	if err != nil {
+		return errors.FromAtomix(err)
+	}
+
+	go func() {
+		for {
+			entry, err := entryStream.Next()
+			if err == io.EOF {
+				log.Warn("Entry stream is closed")
```

Comment:
We need to `close` the channel if we're not going to send any more events otherwise the client will just be hanging indefinitely.